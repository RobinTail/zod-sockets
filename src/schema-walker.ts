import { z } from "zod";

interface VariantDependingProps<U> {
  regular: { next: (schema: z.ZodTypeAny) => U };
  each: { prev: U };
  last: {};
}

export type HandlingVariant = keyof VariantDependingProps<unknown>;

type SchemaHandlingProps<
  T extends z.ZodTypeAny,
  U,
  Context extends object,
  Variant extends HandlingVariant,
> = Context &
  VariantDependingProps<U>[Variant] & {
    schema: T;
  };

export type SchemaHandler<
  T extends z.ZodTypeAny,
  U,
  Context extends object = {},
  Variant extends HandlingVariant = "regular",
> = (params: SchemaHandlingProps<T, U, Context, Variant>) => U;

export type HandlingRules<U, Context extends object = {}> = Partial<
  Record<
    z.ZodFirstPartyTypeKind,
    SchemaHandler<any, U, Context> // keeping "any" here in order to avoid excessive complexity
  >
>;

/** @since 10.1.1 calling onEach _after_ handler and giving it the previously achieved result */
export const walkSchema = <U extends object, Context extends object = {}>({
  schema,
  onEach,
  rules,
  onMissing,
  ...rest
}: SchemaHandlingProps<z.ZodTypeAny, U, Context, "last"> & {
  onEach?: SchemaHandler<z.ZodTypeAny, U, Context, "each">;
  rules: HandlingRules<U, Context>;
  onMissing: SchemaHandler<z.ZodTypeAny, U, Context, "last">;
}): U => {
  const handler = rules[schema._def.typeName as keyof typeof rules];
  const ctx = rest as unknown as Context;
  const next = (subject: z.ZodTypeAny) =>
    walkSchema({ schema: subject, ...ctx, onEach, rules, onMissing });
  const result = handler
    ? handler({ schema, ...ctx, next })
    : onMissing({ schema, ...ctx });
  const overrides = onEach && onEach({ schema, prev: result, ...ctx });
  return overrides ? { ...result, ...overrides } : result;
};
