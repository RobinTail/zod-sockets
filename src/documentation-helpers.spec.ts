import assert from "node:assert/strict";
import { z } from "zod";
import {
  AsyncAPIContext,
  depictAny,
  depictArray,
  depictBigInt,
  depictBoolean,
  depictBranded,
  depictCatch,
  depictDate,
  depictDefault,
  depictDiscriminatedUnion,
  depictEffect,
  depictEnum,
  depictIntersection,
  depictLiteral,
  depictNull,
  depictNullable,
  depictNumber,
  depictObject,
  depictObjectProperties,
  depictOptional,
  depictPipeline,
  depictReadonly,
  depictRecord,
  depictString,
  depictTuple,
  depictUnion,
  depicters,
  onEach,
  onMissing,
} from "./documentation-helpers";
import { walkSchema } from "./schema-walker";
import { describe, expect, test } from "vitest";

describe("Documentation helpers", () => {
  const requestCtx: AsyncAPIContext = {
    direction: "in",
  };
  const responseCtx: AsyncAPIContext = {
    direction: "out",
  };
  const makeNext = (ctx: AsyncAPIContext) => (schema: z.ZodTypeAny) =>
    walkSchema({
      schema,
      rules: depicters,
      ...ctx,
      onEach,
      onMissing,
    });

  describe("depictDefault()", () => {
    test("should set default property", () => {
      expect(
        depictDefault({
          schema: z.boolean().default(true),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictCatch()", () => {
    test("should pass next depicter", () => {
      expect(
        depictCatch({
          schema: z.boolean().catch(true),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictAny()", () => {
    test("should set format:any", () => {
      expect(
        depictAny({
          schema: z.any(),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictUnion()", () => {
    test("should wrap next depicters into oneOf property", () => {
      expect(
        depictUnion({
          schema: z.string().or(z.number()),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictDiscriminatedUnion()", () => {
    test("should wrap next depicters in oneOf prop and set discriminator prop", () => {
      expect(
        depictDiscriminatedUnion({
          schema: z.discriminatedUnion("status", [
            z.object({ status: z.literal("success"), data: z.any() }),
            z.object({
              status: z.literal("error"),
              error: z.object({ message: z.string() }),
            }),
          ]),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictIntersection()", () => {
    test("should flatten two object schemas", () => {
      expect(
        depictIntersection({
          schema: z
            .object({ one: z.number() })
            .and(z.object({ two: z.number() })),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });

    test("should merge deeply", () => {
      expect(
        depictIntersection({
          schema: z
            .object({ test: z.object({ a: z.number() }) })
            .and(z.object({ test: z.object({ b: z.number() }) })),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });

    test("should flatten three object schemas", () => {
      expect(
        depictIntersection({
          schema: z
            .object({ one: z.number() })
            .and(z.object({ two: z.number() }))
            .and(z.object({ three: z.number() })),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });

    test("should maintain uniqueness in the array of required props", () => {
      expect(
        depictIntersection({
          schema: z
            .record(z.literal("test"), z.number())
            .and(z.object({ test: z.literal(5) })),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });

    test.each([
      z.record(z.string(), z.number()).and(z.object({ test: z.number() })), // has additionalProperties
      z.number().and(z.literal(5)), // not objects
    ])("should fall back to allOf in other cases %#", (schema) => {
      expect(
        depictIntersection({
          schema,
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictOptional()", () => {
    test.each<AsyncAPIContext>([requestCtx, responseCtx])(
      "should pass the next depicter %#",
      (ctx) => {
        expect(
          depictOptional({
            schema: z.string().optional(),
            ...ctx,
            next: makeNext(ctx),
          }),
        ).toMatchSnapshot();
      },
    );
  });

  describe("depictNullable()", () => {
    test.each<AsyncAPIContext>([requestCtx, responseCtx])(
      "should add null to the type %#",
      (ctx) => {
        expect(
          depictNullable({
            schema: z.string().nullable(),
            ...ctx,
            next: makeNext(ctx),
          }),
        ).toMatchSnapshot();
      },
    );

    test.each([
      z.string().nullable(),
      z.null().nullable(),
      z.string().nullable().nullable(),
    ])("should only add null type once %#", (schema) => {
      expect(
        depictNullable({
          schema,
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictEnum()", () => {
    enum Test {
      one = "ONE",
      two = "TWO",
    }
    test.each([z.enum(["one", "two"]), z.nativeEnum(Test)])(
      "should set type and enum properties",
      (schema) => {
        expect(
          depictEnum({
            schema,
            ...requestCtx,
            next: makeNext(requestCtx),
          }),
        ).toMatchSnapshot();
      },
    );
  });

  describe("depictLiteral()", () => {
    test("should set type and involve enum property", () => {
      expect(
        depictLiteral({
          schema: z.literal("testing"),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictObject()", () => {
    test.each<{ ctx: AsyncAPIContext; shape: z.ZodRawShape }>([
      { ctx: requestCtx, shape: { a: z.number(), b: z.string() } },
      { ctx: responseCtx, shape: { a: z.number(), b: z.string() } },
      {
        ctx: responseCtx,
        shape: { a: z.coerce.number(), b: z.string({ coerce: true }) },
      },
      { ctx: responseCtx, shape: { a: z.number(), b: z.string().optional() } },
      {
        ctx: requestCtx,
        shape: { a: z.number().optional(), b: z.coerce.string() },
      },
    ])(
      "should type:object, properties and required props %#",
      ({ shape, ctx }) => {
        expect(
          depictObject({
            schema: z.object(shape),
            ...ctx,
            next: makeNext(ctx),
          }),
        ).toMatchSnapshot();
      },
    );

    test("Bug #758", () => {
      const schema = z.object({
        a: z.string(),
        b: z.coerce.string(),
        c: z.coerce.string().optional(),
      });
      expect(
        depictObject({
          schema,
          ...responseCtx,
          next: makeNext(responseCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictNull()", () => {
    test("should give type:null", () => {
      expect(
        depictNull({
          schema: z.null(),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictBoolean()", () => {
    test("should set type:boolean", () => {
      expect(
        depictBoolean({
          schema: z.boolean(),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictBigInt()", () => {
    test("should set type:integer and format:bigint", () => {
      expect(
        depictBigInt({
          schema: z.bigint(),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictRecord()", () => {
    test.each([
      z.record(z.boolean()),
      z.record(z.string(), z.boolean()),
      z.record(z.enum(["one", "two"]), z.boolean()),
      z.record(z.literal("testing"), z.boolean()),
      z.record(z.literal("one").or(z.literal("two")), z.boolean()),
      z.record(z.any()), // Issue #900
    ])(
      "should set properties+required or additionalProperties props %#",
      (schema) => {
        expect(
          depictRecord({
            schema,
            ...requestCtx,
            next: makeNext(requestCtx),
          }),
        ).toMatchSnapshot();
      },
    );
  });

  describe("depictArray()", () => {
    test("should set type:array and pass items depiction", () => {
      expect(
        depictArray({
          schema: z.array(z.boolean()),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictTuple()", () => {
    test("should utilize prefixItems", () => {
      expect(
        depictTuple({
          schema: z.tuple([z.boolean(), z.string(), z.literal("test")]),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictString()", () => {
    test("should set type:string", () => {
      expect(
        depictString({
          schema: z.string(),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });

    test.each([
      z.string().email().min(10).max(20),
      z.string().url().length(15),
      z.string().uuid(),
      z.string().cuid(),
      z.string().datetime(),
      z.string().datetime({ offset: true }),
      z.string().regex(/^\d+.\d+.\d+$/),
    ])("should set format, pattern and min/maxLength props %#", (schema) => {
      expect(
        depictString({
          schema,
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictNumber()", () => {
    test.each([z.number(), z.number().int().min(10).max(20)])(
      "should type:number, min/max, format and exclusiveness props",
      (schema) => {
        expect(
          depictNumber({
            schema,
            ...requestCtx,
            next: makeNext(requestCtx),
          }),
        ).toMatchSnapshot();
      },
    );
  });

  describe("depictObjectProperties()", () => {
    test("should wrap next depicters in a shape of object", () => {
      expect(
        depictObjectProperties({
          schema: z.object({
            one: z.string(),
            two: z.boolean(),
          }),
          ...requestCtx,
          next: makeNext(requestCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictEffect()", () => {
    test.each<{
      ctx: AsyncAPIContext;
      schema: z.ZodEffects<any>;
      expected: string;
    }>([
      {
        schema: z.string().transform((v) => parseInt(v, 10)),
        ctx: responseCtx,
        expected: "number (out)",
      },
      {
        schema: z.string().transform((v) => parseInt(v, 10)),
        ctx: requestCtx,
        expected: "string (in)",
      },
      {
        schema: z.preprocess((v) => parseInt(`${v}`, 10), z.string()),
        ctx: requestCtx,
        expected: "string (preprocess)",
      },
      {
        schema: z
          .object({ s: z.string() })
          .refine(() => false, { message: "test" }),
        ctx: requestCtx,
        expected: "object (refinement)",
      },
    ])("should depict as $expected", ({ schema, ctx }) => {
      expect(
        depictEffect({
          schema,
          ...ctx,
          next: makeNext(ctx),
        }),
      ).toMatchSnapshot();
    });

    test.each([
      z.number().transform((num) => () => num),
      z.number().transform(() => assert.fail("this should be handled")),
    ])("should handle edge cases", (schema) => {
      expect(
        depictEffect({
          schema,
          ...responseCtx,
          next: makeNext(responseCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictPipeline", () => {
    test.each<{ ctx: AsyncAPIContext; expected: string }>([
      { ctx: responseCtx, expected: "boolean (out)" },
      { ctx: requestCtx, expected: "string (in)" },
    ])("should depict as $expected", ({ ctx }) => {
      expect(
        depictPipeline({
          schema: z.string().pipe(z.coerce.boolean()),
          ...ctx,
          next: makeNext(ctx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictDate", () => {
    test.each<AsyncAPIContext>([responseCtx, requestCtx])(
      "should set format date %#",
      (ctx) => {
        expect(
          depictDate({
            schema: z.date(),
            ...ctx,
            next: makeNext(ctx),
          }),
        ).toMatchSnapshot();
      },
    );
  });

  describe("depictBranded", () => {
    test("should pass the next depicter", () => {
      expect(
        depictBranded({
          schema: z.string().min(2).brand<"Test">(),
          ...responseCtx,
          next: makeNext(responseCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("depictReadonly", () => {
    test("should pass the next depicter", () => {
      expect(
        depictReadonly({
          schema: z.string().readonly(),
          ...responseCtx,
          next: makeNext(responseCtx),
        }),
      ).toMatchSnapshot();
    });
  });

  describe("onEach()", () => {
    test.each([requestCtx, responseCtx])(
      "should skip reference objects",
      (ctx) => {
        expect(
          onEach({ prev: { $ref: "some ref" }, ...ctx, schema: z.tuple([]) }),
        ).toEqual({});
      },
    );
  });
});
