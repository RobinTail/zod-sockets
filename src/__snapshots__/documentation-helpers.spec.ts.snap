// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Documentation helpers > depictAny() > should set format:any 1`] = `
{
  "format": "any",
}
`;

exports[`Documentation helpers > depictArray() > should set type:array and pass items depiction 1`] = `
{
  "items": {
    "type": "boolean",
  },
  "type": "array",
}
`;

exports[`Documentation helpers > depictBigInt() > should set type:integer and format:bigint 1`] = `
{
  "format": "bigint",
  "type": "integer",
}
`;

exports[`Documentation helpers > depictBoolean() > should set type:boolean 1`] = `
{
  "type": "boolean",
}
`;

exports[`Documentation helpers > depictBranded > should pass the next depicter 1`] = `
{
  "minLength": 2,
  "type": "string",
}
`;

exports[`Documentation helpers > depictCatch() > should pass next depicter 1`] = `
{
  "type": "boolean",
}
`;

exports[`Documentation helpers > depictDate > should set format date 0 1`] = `
{
  "format": "date",
}
`;

exports[`Documentation helpers > depictDate > should set format date 1 1`] = `
{
  "format": "date",
}
`;

exports[`Documentation helpers > depictDefault() > should set default property 1`] = `
{
  "default": true,
  "type": "boolean",
}
`;

exports[`Documentation helpers > depictDiscriminatedUnion() > should wrap next depicters in oneOf prop and set discriminator prop 1`] = `
{
  "discriminator": {
    "propertyName": "status",
  },
  "oneOf": [
    {
      "properties": {
        "data": {
          "format": "any",
        },
        "status": {
          "enum": [
            "success",
          ],
          "type": "string",
        },
      },
      "required": [
        "status",
      ],
      "type": "object",
    },
    {
      "properties": {
        "error": {
          "properties": {
            "message": {
              "type": "string",
            },
          },
          "required": [
            "message",
          ],
          "type": "object",
        },
        "status": {
          "enum": [
            "error",
          ],
          "type": "string",
        },
      },
      "required": [
        "status",
        "error",
      ],
      "type": "object",
    },
  ],
}
`;

exports[`Documentation helpers > depictEffect() > should depict as 'number (out)' 1`] = `
{
  "type": "number",
}
`;

exports[`Documentation helpers > depictEffect() > should depict as 'object (refinement)' 1`] = `
{
  "properties": {
    "s": {
      "type": "string",
    },
  },
  "required": [
    "s",
  ],
  "type": "object",
}
`;

exports[`Documentation helpers > depictEffect() > should depict as 'string (in)' 1`] = `
{
  "type": "string",
}
`;

exports[`Documentation helpers > depictEffect() > should depict as 'string (preprocess)' 1`] = `
{
  "format": "string (preprocessed)",
}
`;

exports[`Documentation helpers > depictEffect() > should handle edge cases 1`] = `
{
  "format": "any",
}
`;

exports[`Documentation helpers > depictEffect() > should handle edge cases 2`] = `
{
  "format": "any",
}
`;

exports[`Documentation helpers > depictEnum() > should set type and enum properties 1`] = `
{
  "enum": [
    "one",
    "two",
  ],
  "type": "string",
}
`;

exports[`Documentation helpers > depictEnum() > should set type and enum properties 2`] = `
{
  "enum": [
    "ONE",
    "TWO",
  ],
  "type": "string",
}
`;

exports[`Documentation helpers > depictIntersection() > should fall back to allOf in other cases 0 1`] = `
{
  "allOf": [
    {
      "additionalProperties": {
        "format": "double",
        "maximum": 1.7976931348623157e+308,
        "minimum": -1.7976931348623157e+308,
        "type": "number",
      },
      "type": "object",
    },
    {
      "properties": {
        "test": {
          "format": "double",
          "maximum": 1.7976931348623157e+308,
          "minimum": -1.7976931348623157e+308,
          "type": "number",
        },
      },
      "required": [
        "test",
      ],
      "type": "object",
    },
  ],
}
`;

exports[`Documentation helpers > depictIntersection() > should fall back to allOf in other cases 1 1`] = `
{
  "allOf": [
    {
      "format": "double",
      "maximum": 1.7976931348623157e+308,
      "minimum": -1.7976931348623157e+308,
      "type": "number",
    },
    {
      "enum": [
        5,
      ],
      "type": "number",
    },
  ],
}
`;

exports[`Documentation helpers > depictIntersection() > should flatten three object schemas 1`] = `
{
  "properties": {
    "one": {
      "format": "double",
      "maximum": 1.7976931348623157e+308,
      "minimum": -1.7976931348623157e+308,
      "type": "number",
    },
    "three": {
      "format": "double",
      "maximum": 1.7976931348623157e+308,
      "minimum": -1.7976931348623157e+308,
      "type": "number",
    },
    "two": {
      "format": "double",
      "maximum": 1.7976931348623157e+308,
      "minimum": -1.7976931348623157e+308,
      "type": "number",
    },
  },
  "required": [
    "one",
    "two",
    "three",
  ],
  "type": "object",
}
`;

exports[`Documentation helpers > depictIntersection() > should flatten two object schemas 1`] = `
{
  "properties": {
    "one": {
      "format": "double",
      "maximum": 1.7976931348623157e+308,
      "minimum": -1.7976931348623157e+308,
      "type": "number",
    },
    "two": {
      "format": "double",
      "maximum": 1.7976931348623157e+308,
      "minimum": -1.7976931348623157e+308,
      "type": "number",
    },
  },
  "required": [
    "one",
    "two",
  ],
  "type": "object",
}
`;

exports[`Documentation helpers > depictIntersection() > should maintain uniqueness in the array of required props 1`] = `
{
  "properties": {
    "test": {
      "enum": [
        5,
      ],
      "format": "double",
      "maximum": 1.7976931348623157e+308,
      "minimum": -1.7976931348623157e+308,
      "type": "number",
    },
  },
  "required": [
    "test",
  ],
  "type": "object",
}
`;

exports[`Documentation helpers > depictIntersection() > should merge deeply 1`] = `
{
  "properties": {
    "test": {
      "properties": {
        "a": {
          "format": "double",
          "maximum": 1.7976931348623157e+308,
          "minimum": -1.7976931348623157e+308,
          "type": "number",
        },
        "b": {
          "format": "double",
          "maximum": 1.7976931348623157e+308,
          "minimum": -1.7976931348623157e+308,
          "type": "number",
        },
      },
      "required": [
        "a",
        "b",
      ],
      "type": "object",
    },
  },
  "required": [
    "test",
  ],
  "type": "object",
}
`;

exports[`Documentation helpers > depictLiteral() > should set type and involve enum property 1`] = `
{
  "enum": [
    "testing",
  ],
  "type": "string",
}
`;

exports[`Documentation helpers > depictNull() > should give type:null 1`] = `
{
  "type": "null",
}
`;

exports[`Documentation helpers > depictNullable() > should add null to the type 0 1`] = `
{
  "type": [
    "string",
    "null",
  ],
}
`;

exports[`Documentation helpers > depictNullable() > should add null to the type 1 1`] = `
{
  "type": [
    "string",
    "null",
  ],
}
`;

exports[`Documentation helpers > depictNullable() > should only add null type once 0 1`] = `
{
  "type": [
    "string",
    "null",
  ],
}
`;

exports[`Documentation helpers > depictNullable() > should only add null type once 1 1`] = `
{
  "type": [
    "null",
  ],
}
`;

exports[`Documentation helpers > depictNullable() > should only add null type once 2 1`] = `
{
  "type": [
    "string",
    "null",
  ],
}
`;

exports[`Documentation helpers > depictNumber() > should type:number, min/max, format and exclusiveness props 1`] = `
{
  "format": "double",
  "maximum": 1.7976931348623157e+308,
  "minimum": -1.7976931348623157e+308,
  "type": "number",
}
`;

exports[`Documentation helpers > depictNumber() > should type:number, min/max, format and exclusiveness props 2`] = `
{
  "format": "int64",
  "maximum": 20,
  "minimum": 10,
  "type": "integer",
}
`;

exports[`Documentation helpers > depictObject() > Bug #758 1`] = `
{
  "properties": {
    "a": {
      "type": "string",
    },
    "b": {
      "type": "string",
    },
    "c": {
      "type": [
        "string",
        "null",
      ],
    },
  },
  "required": [
    "a",
    "b",
  ],
  "type": "object",
}
`;

exports[`Documentation helpers > depictObject() > should type:object, properties and required props 0 1`] = `
{
  "properties": {
    "a": {
      "format": "double",
      "maximum": 1.7976931348623157e+308,
      "minimum": -1.7976931348623157e+308,
      "type": "number",
    },
    "b": {
      "type": "string",
    },
  },
  "required": [
    "a",
    "b",
  ],
  "type": "object",
}
`;

exports[`Documentation helpers > depictObject() > should type:object, properties and required props 1 1`] = `
{
  "properties": {
    "a": {
      "format": "double",
      "maximum": 1.7976931348623157e+308,
      "minimum": -1.7976931348623157e+308,
      "type": "number",
    },
    "b": {
      "type": "string",
    },
  },
  "required": [
    "a",
    "b",
  ],
  "type": "object",
}
`;

exports[`Documentation helpers > depictObject() > should type:object, properties and required props 2 1`] = `
{
  "properties": {
    "a": {
      "format": "double",
      "maximum": 1.7976931348623157e+308,
      "minimum": -1.7976931348623157e+308,
      "type": "number",
    },
    "b": {
      "type": "string",
    },
  },
  "required": [
    "a",
    "b",
  ],
  "type": "object",
}
`;

exports[`Documentation helpers > depictObject() > should type:object, properties and required props 3 1`] = `
{
  "properties": {
    "a": {
      "format": "double",
      "maximum": 1.7976931348623157e+308,
      "minimum": -1.7976931348623157e+308,
      "type": "number",
    },
    "b": {
      "type": "string",
    },
  },
  "required": [
    "a",
  ],
  "type": "object",
}
`;

exports[`Documentation helpers > depictObject() > should type:object, properties and required props 4 1`] = `
{
  "properties": {
    "a": {
      "format": "double",
      "maximum": 1.7976931348623157e+308,
      "minimum": -1.7976931348623157e+308,
      "type": "number",
    },
    "b": {
      "type": [
        "string",
        "null",
      ],
    },
  },
  "type": "object",
}
`;

exports[`Documentation helpers > depictObjectProperties() > should wrap next depicters in a shape of object 1`] = `
{
  "one": {
    "type": "string",
  },
  "two": {
    "type": "boolean",
  },
}
`;

exports[`Documentation helpers > depictOptional() > should pass the next depicter 0 1`] = `
{
  "type": "string",
}
`;

exports[`Documentation helpers > depictOptional() > should pass the next depicter 1 1`] = `
{
  "type": "string",
}
`;

exports[`Documentation helpers > depictPipeline > should depict as 'boolean (out)' 1`] = `
{
  "type": "boolean",
}
`;

exports[`Documentation helpers > depictPipeline > should depict as 'string (in)' 1`] = `
{
  "type": "string",
}
`;

exports[`Documentation helpers > depictReadonly > should pass the next depicter 1`] = `
{
  "type": "string",
}
`;

exports[`Documentation helpers > depictRecord() > should set properties+required or additionalProperties props 0 1`] = `
{
  "additionalProperties": {
    "type": "boolean",
  },
  "type": "object",
}
`;

exports[`Documentation helpers > depictRecord() > should set properties+required or additionalProperties props 1 1`] = `
{
  "additionalProperties": {
    "type": "boolean",
  },
  "type": "object",
}
`;

exports[`Documentation helpers > depictRecord() > should set properties+required or additionalProperties props 2 1`] = `
{
  "properties": {
    "one": {
      "type": "boolean",
    },
    "two": {
      "type": "boolean",
    },
  },
  "required": [
    "one",
    "two",
  ],
  "type": "object",
}
`;

exports[`Documentation helpers > depictRecord() > should set properties+required or additionalProperties props 3 1`] = `
{
  "properties": {
    "testing": {
      "type": "boolean",
    },
  },
  "required": [
    "testing",
  ],
  "type": "object",
}
`;

exports[`Documentation helpers > depictRecord() > should set properties+required or additionalProperties props 4 1`] = `
{
  "properties": {
    "one": {
      "type": "boolean",
    },
    "two": {
      "type": "boolean",
    },
  },
  "required": [
    "one",
    "two",
  ],
  "type": "object",
}
`;

exports[`Documentation helpers > depictRecord() > should set properties+required or additionalProperties props 5 1`] = `
{
  "additionalProperties": {
    "format": "any",
  },
  "type": "object",
}
`;

exports[`Documentation helpers > depictString() > should set format, pattern and min/maxLength props 0 1`] = `
{
  "format": "email",
  "maxLength": 20,
  "minLength": 10,
  "type": "string",
}
`;

exports[`Documentation helpers > depictString() > should set format, pattern and min/maxLength props 1 1`] = `
{
  "format": "url",
  "type": "string",
}
`;

exports[`Documentation helpers > depictString() > should set format, pattern and min/maxLength props 2 1`] = `
{
  "format": "uuid",
  "type": "string",
}
`;

exports[`Documentation helpers > depictString() > should set format, pattern and min/maxLength props 3 1`] = `
{
  "format": "cuid",
  "type": "string",
}
`;

exports[`Documentation helpers > depictString() > should set format, pattern and min/maxLength props 4 1`] = `
{
  "format": "date-time",
  "pattern": "^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\.\\d+)?Z$",
  "type": "string",
}
`;

exports[`Documentation helpers > depictString() > should set format, pattern and min/maxLength props 5 1`] = `
{
  "format": "date-time",
  "pattern": "^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\.\\d+)?(([+-]\\d{2}:\\d{2})|Z)$",
  "type": "string",
}
`;

exports[`Documentation helpers > depictString() > should set format, pattern and min/maxLength props 6 1`] = `
{
  "pattern": "^\\d+.\\d+.\\d+$",
  "type": "string",
}
`;

exports[`Documentation helpers > depictString() > should set type:string 1`] = `
{
  "type": "string",
}
`;

exports[`Documentation helpers > depictTuple() > should depict as an object with numeric keys 1`] = `
{
  "additionalProperties": false,
  "format": "tuple",
  "properties": {
    "0": {
      "type": "boolean",
    },
    "1": {
      "type": "string",
    },
    "2": {
      "enum": [
        "test",
      ],
      "type": "string",
    },
  },
  "required": [
    "0",
    "1",
    "2",
  ],
  "type": "object",
}
`;

exports[`Documentation helpers > depictUnion() > should wrap next depicters into oneOf property 1`] = `
{
  "oneOf": [
    {
      "type": "string",
    },
    {
      "format": "double",
      "maximum": 1.7976931348623157e+308,
      "minimum": -1.7976931348623157e+308,
      "type": "number",
    },
  ],
}
`;
